crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -248,59 +156,46 @@ impl IndexAllocator {
-                    //
-                    // Note that we do this to maintain an unbiased stealing
-                    // distribution: we want the likelihood of our taking a slot
-                    // from some other module's freelist to be proportional to
-                    // that module's freelist length. Or in other words, every
-                    // *slot* should be equally likely to be stolen. The
-                    // alternative, where we pick the victim module freelist
-                    // first, means that either a module with an affinity
-                    // freelist of one slot has the same chances of losing that
-                    // slot as one with a hundred slots; or else we need a
-                    // weighted random choice among modules, which is just as
-                    // complex as this process.
-                    //
-                    // We don't bother picking an empty slot (no established
-                    // affinity) before a random slot, because this is more
-                    // complex, and in the steady state, all slots will see at
-                    // least one instantiation very quickly, so there will never
-                    // (past an initial phase) be a slot with no affinity.
-                    PoolingAllocationStrategy::ReuseAffinity => inner
-                        .pick_affine(module_id)
-                        .or_else(|| inner.pick_random())?,
+        // As a first-pass always attempt an affine allocation. This will
+        // succeed if any slots are considered affine to `module_id` (if it's
+        // specified). Failing that something else is attempted to be chosen.
+        let slot_id = inner.pick_affine(module_id).or_else(|| {
+            match mode {
+                // If any slot is requested then this is a normal instantiation
+                // looking for an index. Without any affine candidates there are
+                // two options here:
+                //
+                // 1. Pick a slot amongst previously allocated slots
+                // 2. Pick a slot that's never been used before
+                //
+                // The choice here is guided by the initial configuration of
+                // `max_unused_warm_slots`. If our unused warm slots, which are
+                // likely all affine, is below this threshold then the affinity
+                // of the warm slots isn't tampered with and first a cold slot
+                // is chosen. If the cold slot allocation fails, however, a warm
+                // slot is evicted.
+                //
+                // The opposite happens when we're above our threshold for the
+                // maximum number of warm slots, meaning that a warm slot is
+                // attempted to be picked from first with a cold slot following
+                // that.
+                AllocMode::AnySlot => {
+                    if inner.unused_warm_slots < inner.max_unused_warm_slots {
```

Comment:
I started off wondering if this should be `<=` instead, and eventually convinced myself that this is correct, but for a pretty subtle reason that I think should be documented (if it's right).

My initial uncertainty came because we aren't creating new unused warm slots here (rather, we're creating an *used* slot, by allocating it). So if our unused-warm count is *at* the maximum (the `=` case), we don't violate our promise by taking a new cold slot.

The subtle bit though is then once this slot is eventually terminated, we're already committed to having a new unused warm slot, unless we actively reclaim down to the limit (which would take more plumbing). So really we're conservatively accounting for that later effect.

Could we add a (short) comment to this effect?

Reply:
Ah so full disclosure I didn't think about `<` or `<=` at all here. I wrote one down, it passed tests, and I never looked back.

To be clear though the `max_unused_warm_slots` isn't a hard contract or anything, it's more of a guideline than anything else. If the max warm slots is set to 5, for example, but you create 10 instances concurrently, then when those 10 instances are deallocated there will be 10 warm slots.

In that sense I don't think that the `<` or `<=` really matters here since off-by-one is expected to happen naturally when the concurrency limit goes high.

Reply:
Yeah, that's a good point, we don't do reclamation so it's really heuristic rather than a true invariant. Seems fine then.